File: 0494-target-sum/0494-target-sum.java
class Solution {
    private int backtracking(int[] nums, int target, int curr, int sum) {
        if (curr == nums.length) {
            if (sum == target) {
                return 1;
            }
        } else {
            return this.backtracking(nums, target, curr+1, sum+nums[curr]) + this.backtracking(nums, target, curr+1, sum - nums[curr]);
        }
        return 0;
    }
    public int findTargetSumWays(int[] nums, int target) {
        return this.backtracking(nums, target, 0, 0);
    }
}

File: 0494-target-sum/0494-target-sum.py
class Solution:
    def backtrack(self, curr: int, nums: List[int], target: int, result: int) -> int:
        if (curr, result) in self.memo:
            return self.memo[(curr, result)]
        if curr == len(nums):
            if result == target:
                return 1
            else:
                return 0
        else:
            positive = self.backtrack(curr+1, nums, target, result + nums[curr]) 
            negative = self.backtrack(curr+1, nums, target, result - nums[curr])
            self.memo[(curr, result)] = positive + negative
            return self.memo[(curr, result)]

    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        self.memo = dict()
        return self.backtrack(0, nums, target, 0)